### ISSUE TITLE
Incorrect result for $transclude.isSlotFilled(name)

### ISSUE THREAD

**Issue by grimly** [angular/angular.js#15560]
***Note*: for support questions, please use one of these channels: https://github.com/angular/angular.js/blob/master/CONTRIBUTING.md#question. This repository's issues are reserved for feature requests and bug reports.**

**Do you want to request a *feature* or report a *bug*?**
This is, IMHO, a bug

**What is the current behavior?**
When a region is removed through a "ng-if" directive, this region is still recognized as filled for the parent element.

**If the current behavior is a bug, please provide the steps to reproduce and if possible a minimal demo of the problem via https://plnkr.co or similar (template: http://plnkr.co/edit/tpl:yBpEi4).**
http://codepen.io/anon/pen/KNOELg?editors=1010
The second h1 should not be drawn.

**What is the expected behavior?**
Comments are not parts of the regions as they do not fill the region tag name criteria. In such conditions, the $transclude.isSlotFilled() function should return false.
If the current behavior is the intended behavior, then we need some method to check again if the region is filled with actual content.

**What is the motivation / use case for changing the behavior?**
You should be able to encapsulate a region with your own markup, shall it be with awesome eye candies, but then not use that eye candy when it's not useful at all.

**Which versions of Angular, and which browser / OS are affected by this issue? Did this work in previous versions of Angular? Please also test with the latest stable and snapshot (https://code.angularjs.org/snapshot/) versions.**
This is using the multi-slot transclusion feature so this bug is only relevant for Angular 1.5+ This has been tested against 1.5.8 (codepen demo above) and 1.6.1 (my own project as I write this bug report)

**Other information (e.g. stacktraces, related issues, suggestions how to fix)**
If this require a new compilation pass to check if the transcluded region is empty or not, we should be given a new function for this check as not everyone would accept such expensive checks.

A possible start of a solution is provided as a second version of the component. The interval and cache is required because an infinite digest loop because of the transclusions if the region does not exists.
This solution is as I give it : 
 * **EXPENSIVE** as a missing region will run the check every digest loop.
 * **INACCURATE** as a change to the scope within the digest loop would make any test of the correction fail.
 * **DEPRECATED** as it uses $$ functions.

**Comment by gkalpak** [comment]
This is definitely not a bug. `isSlotFilled()` retrns whether or not the user provided any content for the slot. What exactly the content was, what directives it has on it (e.g. `ngIf` with falsy condition) and what these directives do (e.g. remove the element from the DOM) is irrelevant.

IMO, it is bad design to (or at least not "the Angular way &trade;") to let the DOM content drive your component's behavior. The model should be the source of truth; if your component needs to change it's behavior based on some condition, that condition should be passed directly (as an expression) to the component and not have the component infer the condition based on the DOM state.

##
I am going to leave this open to see what others think, but this _works as expected_ imo.

**Comment by grimly** [comment]
> The model should be the source of truth; if your component needs to change it's behavior based on some condition, that condition should be passed directly (as an expression) to the component and not have the component infer the condition based on the DOM state.
It is as it's the model that adds or removes my region. What I gave on this codepen was a minimalist example, but I have more complex elements that relies on this optional transclusion.

One for instance is a border layout. It has 4 optional north, east, west and south regions, plus the default content as its center.
This element, as I designed it, includes some margins that would be unwanted if one region would exist but have no element.

An other would be a list. You might want to let the container element display its message when no element are to be displayed, right when a ng-repeat displays no element. Repeating the default message multiple times over the application is not a solution.

**Comment by gkalpak** [comment]
In all the cases you mentioned, you are trying to turn the DOM into your component's input (it doesn't matter whether the DOM is a result of another directive or hard-coded - the result is the same). This is not a good practice imo and not a usecase we should support. 

Actually, I think you are trying to use transclusion's fallback content for something it wasn't designed for. If you want to have custom behavior (e.g. displaying a default message or styling/hiding a part of the UI) based on some condition, then your presentational components should get that condition as an input (instead of trying to infer it based on the transcluded content).

<sub>(Again, this is just my opinion and others might disagree.</sub>

**Comment by grimly** [comment]
@gkalpak : While I may understand your point, I can't give myself a reason why optional and/or fallback transclusion exists if relying on the DOM is forbidden.
I have a special case where I have a list from 4 different sources and I have not to display the container of that list if all 4 sources are empty. I find it hard to make my condition with 4 filtered sources.
I can ensure you I blamed the designers to think about a such \<enter any disgusting word\> page but they want what they want.

I've found a "better", but still not ideal, way to identify if my region is present but it has consequences like my other workaround.
It breaks AngularJS compatibility with IE9-10.
It doubles all tested transclusions as the transclusion system only work with duplications (I have no trouble with it but one might call a service as his controller initialises).

My workaround component is available at this place : https://pastebin.ubuntu.com/23770996/ (sorry for the pastebin, I can't access gist.github.com from my place).

It is efficient at what it does, using mutation observers to lower the number of checks done.

This comes to the conclusion that, unlike the current "isSlotFilled" function, that evaluation is **required** to run the check and this step cannot be synchronous (which makes it harder to use than "isSlotFilled").

Since the strategy to transclusion only works through duplication and compilation of a "frozen template", and we (at least I) want the evaluation to be on a different level than the injection point, we might not find a more suitable solution except including it inside the transclusion function. This last idea could give even worst results.

**As I said, if it costs performances to even include it in the framework, I don't want it in the framework. You may tag it as a "wont-fix" then close it. I would still advice this case in some sort of documentation.**

**Comment by gkalpak** [comment]
> I can't give myself a reason why optional and/or fallback transclusion exists if relying on the DOM is forbidden.

I am not sure why you think optional transclusion is in some way related to DOM access :smiley:
(As I am sure you understand already) the purpose of optional transclusion is to be able to omit certain "optional" parts of a component (e.g. an optional footer in a modal component) and the main purpose of fallback content (afaik) is to avoid boilerplate when you want to use a specific template most of the time, but still be able to overwrite it in some occasions.

What we are debating here is how/when Angular determines if transcluded content has been provided or not. The current implementation determines that, based on the presence or absence of content in the initial HTML template (and correctly so imo). This allows Angular to make its decision at compile time and is more static. What you are proposing (i.e. basing the decision on the (dynamic) state of the HTML _after_ the initial content has been compiled and linked) is inherently less performant, would complicate things and also affect the relative order of operations (e.g. you first need to compile/link the content and then decide whether we need to insert/compile/link the fallback content).

##
I am not saying there are no valid usecases for it, nor am I claiming that the current implementation covers every possible usecase one might desire. I am just saying that it is sufficiently efficient, covers a reasonable amount of common usecases and aligns well with "the Angular way".

I do understand that there might be some less common usecases that the current implementation doesn't serve well, in which case you need to implement custom solutions (and yes...if necessary access the DOM 😛 - it is not forbidden btw; it is just discouraged (most of the time)).

##
In a nutshell, I believe the current implementation is reasonable, supports most common usecases and we should stick to that. For more advanced and rare usecases (such as yours), one would have to implement a custom solution.

I don't think it is more helpful than confusing to mention such usecases in the docs, but if you can come up with a way to improve the documentation of optional slots/fallback content/`isSlotFilled()`, you are more than welcome to submit a PR.

<sub>(Closing as _won't fix_.)</sub>
